File: js/http/server4_template.js
/**
 * Same functionality as server3
 * but this time we use a template to render the html
 * which makes it easier to maintain (separation of concerns)
 */

import http from 'http';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { render } from 'ejs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const PORT = 9003;
const STATE = {};

// Read template file
const templatePath = path.join(__dirname, 'template.html.ejs');
const template = fs.readFileSync(templatePath, 'utf-8');

const server = http.createServer((req, res) => {
    const clientAddr = req.socket.remoteAddress;
    console.log(`Server received ${req.method} from ${clientAddr} for path ${req.url}`);

    if (req.method === 'GET') {
        // Handle GET request - display current state using template
        res.writeHead(200, { 'Content-Type': 'text/html' });
        
        // Render template with data
        const html = render(template, {
            client: clientAddr,
            path: req.url,
            data: STATE
        });
        
        res.end(html);
    } else if (req.method === 'POST') {
        // Handle POST request
        if (req.url === '/api/register') {
            let body = '';
            req.on('data', chunk => {
                body += chunk.toString();
            });

            req.on('end', () => {
                try {
                    const data = JSON.parse(body);
                    if (typeof data !== 'object' || data === null) {
                        res.writeHead(400, { 'Content-Type': 'text/plain' });
                        res.end('Expected a dictionary');
                        return;
                    }

                    // Update state with new data
                    Object.assign(STATE, data);
                    console.log('Server received data:', data);

                    res.writeHead(200, { 'Content-Type': 'application/json' });
                    res.end(JSON.stringify(STATE));
                } catch (error) {
                    res.writeHead(400, { 'Content-Type': 'text/plain' });
                    res.end('Invalid JSON data');
                }
            });
        } else {
            res.writeHead(404);
            res.end();
        }
    } else {
        res.writeHead(405);
        res.end();
    }
});

console.log(`Open this in your browser:\nhttp://localhost:${PORT}/index.html`);
server.listen(PORT); 

File: python/http/server4_template.py
"""
same functions as server3
but this time we use a template to render the html
which makes it easier to maintain (separation of concerns)
"""

# pylint: disable = missing-docstring, invalid-name

from http.server import HTTPServer, BaseHTTPRequestHandler
import json

from jinja2 import Template

STATE = {}

class EchoHttpHandlerWithPOST(BaseHTTPRequestHandler):

    def do_GET(self):
        client_addr = self.client_address
        request_path = self.path

        self.send_response(200)
        self.send_header(b'Content-type', b'text/html')
        self.end_headers()
        # -------
        # this is the part that changes from previous example
        # Assuming we start the server from the repo root
        with open("python/http/template.html.j2", "r") as f:
            template = f.read()
            # the variable names we can use in the template
            env = dict(
                client=client_addr,
                path=request_path,
                data=STATE,
            )
            html = Template(template).render(env)
            self.wfile.write(html.encode())
        # -------


    def do_POST(self):
        client_addr = self.client_address
        print(f"Server received POST from {client_addr} for path {self.path}")
        request_path = self.path
        if request_path == "/api/register":
            # Read the data
            data_size = int(self.headers["Content-Length"])
            self.log_message( f"data_size = {data_size}")
            data_raw = self.rfile.read(data_size).decode()
            data_json = json.loads(data_raw)
            # expect a dicionary
            if not isinstance(data_json, dict):
                self.send_response(400)
                self.send_header('Content-type', 'text/plain')
                self.end_headers()
                self.wfile.write(b"Expected a dictionary")
                return
            for k, v in data_json.items():
                STATE[k] = v
            print(f"Server received data: {data_json}")
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            self.wfile.write(json.dumps(STATE).encode())
        else:
            self.send_response(404)

PORT = 9003
print(f"Open this in your browser:\nhttp://localhost:{PORT}/index.html")

server = HTTPServer(("", PORT), EchoHttpHandlerWithPOST)
server.serve_forever()
